docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+  failure modes (split brain, phantom member) to prevent introducing more failure modes. Sophisticated
+  distributed systems designers can implement more sophisticated solutions than PetSet if necessary -
+  new users should not become vulnerable to additional failure modes through an overly flexible design.
+* **Limited scaling** - While flexible scaling is important for some clusters, other examples of clusters
+  do not change scale without significant external intervention. Human intervention may be required after
+  scaling. Changing scale during cluster operation can lead to split brain in quorum systems. It should be
+  possible to scale easily, but the details of making that safe belong to the pods and image authors.
+* **No generic cluster lifecycle** - Rather than design a general purpose lifecycle for clustered software,
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is important. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+}
+```
+
+Like a replication controller, a PetSet may be targeted by an autoscaler. The PetSet makes no assumptions
```

Comment:
Hrm - Does it?  I think it just means that the identities have a maximum size - that actually helps people.  We could define bounded and unbounded ranges, which for some sets is actually extremely valuable.  The intSet allows scale to say "not scalable to X" which in practice is what you want for something that only has 2-3 modes (3, 5, or 7).

It really depends on how attached people are to their names.  Will people feel better naming their pets?  Will it complicate other things (name substitution gets kind of yucky the longer names are - ultimately you need to have a 'cap that' strategy when base name + suffix gets too long).  Some concrete use cases I can think of:

    stringSet: ["master", "slave1", "slave2", "slave3"] # although what i really want is an unbounded number of slaves
    intRange: [ 3, 5, 7 ] # mongodb in replica set mode
    fixedIdentities: ["master"], intRange: { min: 1, max: 6 } # enforces two or more identities

or something like:

    type Identity struct {
      FixedIdentities []string
      Pattern string
      Min int
      Max int
    }
   
    []Identity{
      {Fixed: []string{"master"}, Required: true},
      {Pattern: "slave-$(count)", Min: 1, Max: 10},
    }
    # allows ["master", "slave-1"] and up to 9 slaves

I don't think complex calculation of identity is hard, but we'd need to think through what the changes allowed are.  You might have to manually delete the orphaned master.